### ISSUE TITLE
pyprocess.sig2oise_ratio(corr) function doesn't accept corrs of numpy dtype='object' as an argument.

### ISSUE THREAD

**Issue by nepomnyi** [OpenPIV/openpiv-python#280]
**Is your feature request related to a problem? Please describe.**

The function `pyprocess.sig2noise_ratio(corr)` yields an error if the argument is a numpy array of `dtype='object'`. If the `dtype` of the argument is not `object`, the function works correctly.

### Here's how to reproduce the error.

_First the code that works fine:_
```
import numpy as np
from openpiv import pyprocess

# create a random array of dtype='float64'
a = np.random.default_rng(42).random((320,64,64)) # the dimensions are the ones that I get in my real code

s2n = pyprocess.sig2noise_ratio(a)
```

_Second, the code that doesn't work:_

```
import numpy as np
from openpiv import pyprocess

# create a random array of dtype='float64'
a = np.random.default_rng(42).random((320,64,64)) # the dimensions are the ones that I get in my real code

# convert the array to dtype='object'
b = a.astype('object')

s2n = pyprocess.sig2noise_ratio(b)
```

This code yields the following error:

```
>>> s2n = pyprocess.sig2noise_ratio(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\nepomnyi\Miniconda3\envs\jumpToOpenPIV0.25.0\Lib\site-packages\openpiv\pyprocess.py", line 535, in sig2noise_ratio
    (peak2_i, peak2_j), corr_max2 = find_second_peak(
                                    ^^^^^^^^^^^^^^^^^
  File "C:\Users\nepomnyi\Miniconda3\envs\jumpToOpenPIV0.25.0\Lib\site-packages\openpiv\pyprocess.py", line 315, in find_second_peak
    (i, j), corr_max2 = find_first_peak(tmp)
                        ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\nepomnyi\Miniconda3\envs\jumpToOpenPIV0.25.0\Lib\site-packages\openpiv\pyprocess.py", line 264, in find_first_peak
    return np.unravel_index(np.argmax(corr), corr.shape), corr.max()
                                                          ^^^^^^^^^^
  File "C:\Users\nepomnyi\Miniconda3\envs\jumpToOpenPIV0.25.0\Lib\site-packages\numpy\ma\core.py", line 5885, in max
    result = self.filled(fill_value).max(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\nepomnyi\Miniconda3\envs\jumpToOpenPIV0.25.0\Lib\site-packages\numpy\core\_methods.py", line 40, in _amax
    return umr_maximum(a, axis, None, out, keepdims, initial, where)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '>=' not supported between instances of 'float' and 'str'
>>>
```

My specifications:

- openpiv v0.24.4b0
- numpy v1.23.5
- python v3.11.0
- conda v22.9.0 (miniconda 3)
- Windows 10 Enterprise, v21H2, OS build 19044.2251, feature experience pack 120.2212.4180
- processor Intel(R) Xeon(R) CPU E3-1231 v3 @ 3.40GHz, 3.40GHz
- RAM 16Gb (15.8Gb usable)

**Describe the solution you'd like**

Add the following code between the lines 298 and 300 in the function pyprocess.sig2noise_ratio(corr):

```
if corr.dtype != 'float64':                      
   corr = corr.astype('float64')
```
                  

**Describe alternatives you've considered**

1. First alternative is leave everything as is. A user who creates a `numpy` array of `dtype='object'` should know what he is doing and convert it to `np.float64` before supplying it to `pyprocess.sig2noise_ratio(corr)`.
2. Second alternative is to mention, in the docstring of  `pyprocess.sig2noise_ratio(corr)`, that the argument must not be of `dtype='object'`

**Additional context**

When I was debugging the error, I understood that the reason `dtype='object'` wouldn't get accepted was due to line 304. The function `corr.view(ma.MaskedArray)` does something to an array of `dtype='object'` that causes the function `numpy.max()` - called on line 315 - to yield the error. If one supplies an array of `dtype='object'` to `numpy.max()`, one won't get any errors, i.e. `numpy.max` works correctly with `dtype='object'`. I wasn't being able to find out what exactly happens to a `dtype='object'` array after `corr.view(ma.MaskedArray)`, because that function is written in C.

**Comment by alexlib** [comment]
Dear @nepomnyi - you arrived at very peculiar situation where you create the corr function by yourself. I think that a better way would be just to add to the documentation that if somebody arrives at this point, then the only way to use the openpiv.sig2noise is to use it as float64. I do not think we need to add a typecheck on it - this function is used millions of times during the PIV analysis, and additional `if` statement is just slowing it down. 

I'm not sure either why would you "build" a new corr and not just use the `corr` that is created by the openpiv-python. What I mean is that you aggregate in that multiple correlations and such an example exists in our repo:

https://github.com/OpenPIV/openpiv-python-examples/blob/main/notebooks/ensemble_correlation.ipynb

**Comment by nepomnyi** [comment]
Hello @alexlib,

yes, sure, as you say. I myself was not sure if it was necessary to introduce the type check. And I totally agree with you on the speed reduction due to the additional `if` statement (though to avoid speed reduction, we could do `try: except: ` instead of `if`).

With regards to your remark about building my own `corr` array, as far as I understand, you went through my code I posted in the Google groups. 

The example you cited is, exactly, the one I was using to write my code. And like you, correctly, noticed, I "aggregated" multiple correlations. To be on the same page, let me explain what I mean by "aggregated". When we have one PIV image pair, we, first, apply `moving_window_array` to the images, second, we apply `fft_correlate_images`. But what if we have several PIV image pairs? In your example, you suggest applying `moving_window_array` and  `fft_correlate_images` to the pairs of images in a `for` loop (see block [6] in you Jupeter notebook example). That's what I mean by "aggregated": how to process more than one PIV pair. 

In my experiment, I have 500 images. Python `for` loops are slow. I was thinking how to apply `moving_window_array` + `fft_correlate_images` to the pairs of images more effectively. That's when I realized OpenPIV functions are not "vectorized". I made a note about that to my list of notes of what I want to introduce to OpenPIV  and started searching for a solution.

I found out that the fastest way to apply the same function - `moving_window_array` + `fft_correlate_images` in my case - to a list of objects is to use parallel programming as described [here](https://stackoverflow.com/a/48374491/10073233).

Once I've implemented that, I realized OpenPIV must have had parallel programming option. But since I already had my code, I didn't spend time on looking into that.

Once I've obtained my correlations, I wanted to group them in groups of N (where N can be changed). I could use `sort` as you used in your example (see block [4]). But I decided to go with `numpy` grouping methods because I thought `numpy` would be faster.

There are two ways to implement grouping in `numpy`: `numpy.array_split` and `numpy.split`. Here's the deal with them. Imagine you have ended up having 5 correlations. You want to rearrange them in groups of 2. You'll get 3 groups: the first two groups will contain 2 correlations each and the last group will contain only 1 correlation. `numpy.array_split` allows such situation, but `numpy.split` doesn't and only allows groups with equal numbers of elements.

Initially, I didn't mind to have the last group with less elements. The problem was that, after the grouping, I combined the groups in another `numpy` array (again, because `numpy` is fast). And `numpy` doesn't like to have elements of different shape. If the elements are of different shape, you must assign `dtype='object'` to the array.

That is how I ended up having `dtype = 'object'` correlations being supplied to `sig2noise_ratio`.

But since then, I have moved to having groups of the same shape and that's not a problem for me anymore.

**Comment by nepomnyi** [comment]
Yes, it can be difficult to understand. Yes, I checked the example. I used that example to build my code. I just didn't use the `sort` function to sort my images. But that's not an important detail as far as I understand.

Thank you for the Zoom call offer. That could be helpful. I can do any time, except for today after 6:30pm and tomorrow after 3pm. The times are in LosAngeles time zone. I can do the weekend if that's more convenient for you.
